de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/annotations/AnnotationManager.java

```diff
@@ -0,0 +1,739 @@
+            });
+        });
+    }
+
+    /**
+     * Moves all annotations for the given file forward by the length of the
+     * removal if they are located behind the removed text. Shortens the
+     * annotations if they partially overlap with the removed text. Removes all
+     * annotations that were completely contained in the removed text.
+     * <p>
+     * If a RangeHighlighter is present, this will all be done automatically by
+     * the internal Intellij logic. This is why
+     * {@link AnnotationManager#updateAnnotationStore(IFile)} should
+     * <b>always</b> be called before an editor is closed to synchronize the
+     * boundaries of the saved annotations with the currently displayed
+     * boundaries in the editor annotation model and remove invalid
+     * annotations from the annotation store.
+     * </p>
+     * <p></p>
+     * This method should be used to adjust the position of all annotations
+     * after text was added to a currently closed file.
+     *
+     * @param file  the file text was removed from
+     * @param start the start position of removed text
+     * @param end   the end position of the removed text
+     */
+    public void moveAnnotationsAfterDeletion(
+        @NotNull
+            IFile file, int start, int end) {
+
+        if (start == end) {
+            return;
+        }
+
+        checkRange(start, end);
+
+        List<SelectionAnnotation> emptySelectionAnnotations = moveAnnotationsAfterDeletion(
+            selectionAnnotationStore.getAnnotations(file), start, end);
+
+        emptySelectionAnnotations
+            .forEach(selectionAnnotationStore::removeAnnotation);
+
+        List<ContributionAnnotation> emptyContributionAnnotations = moveAnnotationsAfterDeletion(
+            contributionAnnotationQueue.getAnnotations(file), start, end);
+
+        emptyContributionAnnotations
+            .forEach(contributionAnnotationQueue::removeAnnotation);
+    }
+
+    /**
+     * <p>
+     * If there are not range highlighters or editors present:
+     * </p>
+     * Moves all given annotations for the given file forward by the length of
+     * the removal if they are located behind the removed text. Shortens the
+     * annotations if they partially overlap with the removed text. Returns a
+     * list of annotations that were completely contained in the removed text.
+     * <p></p>
+     * Does nothing if the annotation has a local representation (an editor or
+     * range highlighters).
+     *
+     * @param annotations   the annotations to adjust
+     * @param deletionStart the start position of the deleted text
+     * @param deletionEnd   the end position of the deleted text
+     * @param <E>           the annotation type
+     * @return the list of deleted annotations
+     * @see #moveAnnotationsAfterDeletion(IFile, int, int)
+     */
+    @NotNull
+    private <E extends AbstractEditorAnnotation> List<E> moveAnnotationsAfterDeletion(
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 23 to the 15 allowed. ([squid:S3776](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=238&issues=AWYa7GVgvOWp7Yd62c1S&open=AWYa7GVgvOWp7Yd62c1S)

[ISSUE_KEY]: AWYa7GVgvOWp7Yd62c1S